mamba/venom: increase kernel partition size

SkewedZeppelin opened the pull request:
This merge request contains four commits:
1. Increases mamba kernel partition from 3MB to 4MB
2. Increases venom kernel partition from 3MB to 6MB.
This is accomplished by shrinking their root partitions by 1MB and 3MB respectively.
3. Re-enables these devices for buildbots
4. Adds a compatibility version check and message

3MB is not enough for building with many kernel modules included and is not enough for Linux 5.10.
Both of their uboots from factory load up to those respective values.

I have tested mamba working with a 3.2MB 5.4 kernel.
I have tested mamba working with a 3.1MB 5.10 kernel.
@anomeome has tested mamba with a 3.2MB 5.10 kernel.
@solidus1983 has tested venom with a <3MB 5.4 kernel.

Compatibility can be checked like so:
mamba: `fw_printenv | grep "pri_kern_size";` *must* equal 0x400000
venom: `fw_printenv | grep "priKernSize";` *must* equal 0x0600000

Flashing process:
A factory style image is necessary when flashing to or away from a resized build
- backup
- flash a factory style image via sysupgrade, with `force` set and `keep settings` unset
- restore backup
- afterwards, you can flash normally.

Flashing from normal OpenWrt to resized has been tested to work.
Flashing from resized OpenWrt to normal has been tested to work.
Flashing from OEM firmware to resized has been tested to work (via user report).
Flashing from resized to OEM firmware has been tested to work (via user report).

There is more information in the forum thread post:
https://forum.openwrt.org/t/increasing-mamba-and-venom-kernel-partition-to-6mb/87807

SkewedZeppelin committed:
mvebu: mamba resize kernel to 4MB

mamba has a 3MB kernel partition as specified by the DTS.
3MB is not sufficient for building with many kernel modules or newer
kernel versions.

mamba uboot however as set from factory will load up to 4MB.
This can be observed by looking a uboot log:
	NAND read: device 0 offset 0xa00000, size 0x400000
	4194304 bytes read: OK
and from uboot environment variables:
	$ fw_printenv | grep "pri_kern_size";
	pri_kern_size=0x400000

Resize the root partitions from 37MB to 36MB to let kernel expand
into it another 1MB.
And set kernel target size to 4MB.

Lastly add a compatibility version message: kernel-size-migration.
And set it on mamba to prevent sysupgrading without first reinstalling from
a factory image.

Signed-off-by: Tad Davanzo <tad@spotco.us>

SkewedZeppelin committed:
mvebu: venom resize kernel to 6MB

venom has a 3MB kernel partition as specified by the DTS.
3MB is not sufficient for building with many kernel modules or newer
kernel versions.

venom uboot however as set from factory will load up to 6MB.
This can be observed by looking a uboot log:
	NAND read: device 0 offset 0x900000, size 0x600000
	6291456 bytes read: OK
and from uboot environment variables:
	$ fw_printenv | grep "priKernSize";
	priKernSize=0x0600000

Resize the root partitions from 120MB to 117MB to let kernel expand
into it another 3MB.
And set kernel target size to 6MB.

Lastly set the kernel-size-migration compatibility version on venom to
prevent sysupgrading without first reinstalling from a factory image.

Signed-off-by: Tad Davanzo <tad@spotco.us>

adschm commented:
Please update the Signed-off-by with your full real name.

adschm commented:
Be aware that buildbot images will be bigger, so final number for kernel sizes should be acquired with buildbot settings and all feeds installed (unless you did that already, but number reported seem to low from my experience).

SkewedZeppelin commented:
@adschm 

I've updated the Signed-off-by, GitHub seems to be slow.

> buildbot images will be bigger

I will test this soon.

SkewedZeppelin commented:
Building 5.4 with
```
CONFIG_ALL_KMODS=y
CONFIG_ALL_NONSHARED=y
CONFIG_STRIP_KERNEL_EXPORTS=n
CONFIG_KERNEL_CC_OPTIMIZE_FOR_PERFORMANCE=y
```
yields a 3,203,306 byte kernel.

Tested booting on mamba.

Current OpenWrt snapshot for WRT3200ACM kernel size is 3,208,600.

SkewedZeppelin commented:
I pushed again and GitHub seems to have finally updated the PR.

Should I create a patch that undoes a1eb2c463352d05161a756f806e8e8a039f4490c and ac9730c49527609e29d87e5c8600ad7313c99c19, or should I create a revert for each of them?

Should I also create a patch for a 1.2 compatibility version? Like 494f12c52df6767ec0fabf2b2fac8f453323a4c5

adschm commented:
> Building 5.4 with
> 
> ```
> CONFIG_ALL_KMODS=y
> CONFIG_ALL_NONSHARED=y
> CONFIG_STRIP_KERNEL_EXPORTS=n
> CONFIG_KERNEL_CC_OPTIMIZE_FOR_PERFORMANCE=y
> ```
> 
> yields a 3,203,306 byte kernel.
> 
> Tested booting on mamba.
> 
> Current OpenWrt snapshot for WRT3200ACM kernel size is 3,208,600.

With all feeds installed?

./scripts/feeds update -a
./scripts/feeds install -a

adschm commented:
> Should I create a patch that undoes a1eb2c4 and ac9730c, or should I create a revert for each of them?

Just enable them in a patch, don't think you need to refer to the older ones for that.

> Should I also create a patch for a 1.2 compatibility version? Like 494f12c

If sysupgrade still works, yes. If reflashing is required, the version should be bumped to 2.0.
In the latter case, you would not use board.d, but uci-defaults like in this case:
https://github.com/openwrt/openwrt/commit/07aa858a73e6e855fc62a37ae275518fa4db5e50
(relevant changes are at the bottom)

adschm commented:
And please do all the changes in one commit per device, i.e. adjust partitions, remove DEFAULT=n, adjust compat version.

SkewedZeppelin committed:
mvebu: enable WRT1900AC v1 and WRT32X for buildbots

Kernel size limits have been dealt with.
Effective revert of a1eb2c46 and ac9730c4.

Signed-off-by: Tad Davanzo <tad@spotco.us>

SkewedZeppelin commented:
Done.
It doesn't seem like compatibility messages stack? Might causes issues for those migrating from swconfig?

solidus1983 commented:
So far no issues after flashing a resized image. Hopefully all goes well on the buildbots.

anomeome commented:
Tested the latest push:

- `option compat_version '2.0'` to only the 2 targets of interest?
- patch version number on mamba target, 3xx , 4xx?
- backup restore of config deserves some elaboration, as you will get a message on the second flash due to the compat still being 1.1 post restore, until you do the second upgrade

adschm commented on target/linux/mvebu/image/cortexa9.mk:
```diff
@@ -112,6 +118,7 @@ TARGET_DEVICES += linksys_wrt1900acs
 define Device/linksys_wrt1900ac-v1
   $(call Device/linksys)
   $(Device/dsa-migration)
```
dsa-migration can obviously be dropped now, as it's the same two variables you reset with the new include.

adschm commented:
Please put the relevant changes directly in the first two patches; they belong together.

> backup restore of config deserves some elaboration, as you will get a message on the second flash due to the compat still being 1.1 post restore, until you do the second upgrade

Well, essentially the compat_version is a property of the config; just consider swconfig vs. DSA. For backup/restore, you would have to change the uci config value manually similar to what uci-defaults does.

Still, in this very case we are left with the strange situation that /etc/board.json and /etc/config/system will have different explicit version numbers stored. I will have to think about that ...

SkewedZeppelin commented:
> dsa-migration can obviously be dropped now

Updated

> Please put the relevant changes directly in the first two patches

To confirm, just two commits total?

nitroshift reviewed (approved):
Run-tested on 2 WRT1900AC v1 (mamba). Confirmed as working.

nitroshift

SkewedZeppelin commented:
@adschm 
Kindly, what changes do I need to make to see this merged?

hauke commented:
Please use the same name as commit author and in the Signed-off-by line.
You can change it with git commit --author="<author name>" or fix it in your ~/.gitconfig

It is good that you do this in separate commits for venom and mamba.
Could you add the call to ```$(Device/kernel-size-migration)``` in the commit which changes the dts file. You can add the ```define Device/kernel-size-migration``` in the commit which changes the dts file for the first device.

SkewedZeppelin commented:
@hauke

I have made the changes you requested.

hauke commented:
Thank you for the pull request, I pushed it to master.

dominick-han commented:
Thanks everyone for making this patch! Finally can run clean no-patch truck builds on my mamba :)
